### PR TITLE
Improve design of options and queue

### PR DESCRIPTION
**Pull request by groenroos**
Improves the UI/UX design of the queue overlay, share dialog, and options screen.

- Designed to more closely mimic Nebula's design:

Current | New
-|-
Screenshot 2021-09-11 at 17 39 14 | Screenshot 2021-09-11 at 17 38 41
Screenshot 2021-09-11 at 17 39 51 | Screenshot 2021-09-11 at 17 41 02
Screenshot 2021-09-11 at 17 48 41 | Screenshot 2021-09-11 at 17 48 25

- Options screen now respects light and dark mode preferences:

Screenshot 2021-09-11 at 17 40 43

**Notes:**
- Nebula's font Whitney is proprietary, Lato from Google Fonts is the closest replacement
- Needed some minor adjustments to some strings, meaning the German translation needs updating/verifying.

### CODE REVIEW

**Comment by cpiber** [comment]
Looks very neat, I'll look at it on Monday when I'm home again!

File: .gitignore [hunk 1]
```diff
@@ -2,6 +2,7 @@ extension-dist
 web-ext-artifacts
 todo.md
 .vscode
+build/
```
Comment (cpiber) [inline .gitignore:5]: Why do you need that? You didn't change anything in the build tooling, so there should not be a build folder

**Review by cpiber — COMMENTED** [review]
I love the new design!

A few minor things:
- The buttons look like they should react to hover (fill with border color), what do you think of that? I'd also like a little feedback for the links
- For the text controls, you let the label become bigger with an empty control on this nice animation, this should however not happen when there is a placeholder, as they then overlap
- On install, the changelog modal displays a little message, you don't account for that, it's glued to the title of the first version, if you could add a spacer there as well would be perfect
image
- Please take a look again at the build-in options in both Firefox and Chrome, the title there looks strange. Previously I had only 'Options' there for exactly that

**Comment by cpiber** [comment]
And could you please merge in [groenroos/feature/design](https://github.com/cpiber/NebulaEnhance/tree/groenroos/feature/design)? It's a branch based off yours with a fix for the CI

File: .gitignore (hunk 1 above)
Comment (groenroos) [reply .gitignore:5]: I indeed didn't change the build tooling, but I did initially struggle to get the build working - at some point, it generated a `build/` folder, which I assumed probably should not be committed.

**Comment by groenroos** [comment]
> I love the new design!

🎉

> * The buttons look like they should react to hover (fill with border color), what do you think of that? I'd also like a little feedback for the links and the input fields

I definitely agree, but I modeled everything about the design based on Nebula's design, and their buttons (e.g. on [the account page](https://beta.nebula.app/account)) do not have a hover state.

> * For the text controls, you let the label become bigger with an empty control on this nice animation, this should however not happen when there is a placeholder, as they then overlap

The tricky bit is that this type of transition is usually controlled by JavaScript (and it is, on Nebula) - but in the absence of that, I did a hack to do this in pure CSS:

https://github.com/cpiber/NebulaEnhance/blob/093905f40468ffd5df2106ce9934c913358116ee/src/styles/helpers/shared.scss#L60-L63

The`:not(:placeholder-shown)` is a proxy for "enlarge the label on blur only if the input doesn't have a value" - based on [this technique](https://stackoverflow.com/a/35302732/951402). To work, it requires the input to have a placeholder containing only space:

https://github.com/cpiber/NebulaEnhance/blob/093905f40468ffd5df2106ce9934c913358116ee/src/scripts/content/queue/html.ts#L35

But as you say, it does break down with inputs that have a genuine placeholder.

- **Easy solution:** Replace all placeholders with `" "`
- **Correct solution:** Replace the CSS hack with proper JS to control the animation of each `.enhancer-field input`

Unfortunately I don't think I have the time anytime soon to write this JS that would simultaneously work both on the options and on the Nebula side.

> * On install, the changelog modal displays a little message, you don't account for that, it's glued to the title of the first version, if you could add a spacer there as well would be perfect

Will do!

> * Please take a look again at the build-in options in both Firefox and Chrome, the title there looks strange. Previously I had only 'Options' there for exactly that

Good point, will do!

**Comment by cpiber** [comment]
> I definitely agree, but I modeled everything about the design based on Nebula's design, and their buttons (e.g. on the account page) do not have a hover state.

You are right! In that case let's leave it as is.

> But as you say, it does break down with inputs that have a genuine placeholder.

Could you not simply prevent it going down if it has a placeholder (`[placeholder]`)? In any case, all inputs have a placeholder, so I don't think it's even necessary

Thank you for your continued work!

**Comment by cpiber** [comment]
Not to be pushy, but when do you think you can finish this? I'd like to include it in the next update if possible

**Comment by groenroos** [comment]
> Not to be pushy, but when do you think you can finish this? I'd like to include it in the next update if possible

No worries! I'll get this ready over the weekend, if that's okay!

**Comment by cpiber** [comment]
Yes, that's perfect, thank you!

**Comment by cpiber** [comment]
I plan on releasing the next version today evening (UTC). I'll include this PR in the next version.

By the way, I will resolve the merge conflict; you don't have to merge master.

**Comment by groenroos** [comment]
@cpiber Hi - sorry for the delay! I've committed fixes for the welcome banner and the modalised settings page, and I converted the label shrinking from the dirty CSS hack to a proper JS/TS solution. Have a look!

**Review by cpiber — CHANGES_REQUESTED** [review]

File: src/scripts/options/form.ts [hunk 2]
```diff
@@ -48,4 +48,19 @@ Array.from(document.querySelectorAll<HTMLElement>('[data-i18n]')).forEach(e => {
   e.dataset.i18n.split(',').forEach(attr =>
     e.setAttribute(attr, e.getAttribute(attr).replace(/__MSG_(.+)__/g, (...args) => getBrowserInstance().i18n.getMessage(args[1]))));
   e.removeAttribute('data-i18n');
-});
\ No newline at end of file
+});
+
+// label animation
+const setInputClass = (el: HTMLInputElement) => {
+  if(el.value === '' || !el.value) {
```
Comment (cpiber) [inline src/scripts/options/form.ts:55]:
```suggestion
  if(!el.value) {
```
is enough

File: src/scripts/options/settings.ts [hunk 3]
```diff
@@ -38,6 +38,9 @@ export const toData = (useDefaults = false) => {
     } else if (els[key].type === 'checkbox') {
       val = !useDefaults ? (els[key] as HTMLInputElement).checked : !!val;
     }
+    if (els[key].classList.contains('enhancer-text-input') && val) {
+      els[key].classList.add('has-value');
+    }
```
Comment (cpiber) [inline src/scripts/options/settings.ts:43]: You need to remove the class again:
```suggestion
    if (els[key].classList.contains('enhancer-text-input')) {
      els[key].classList.toggle('has-value', !!val);
    }
```

**Review by cpiber — CHANGES_REQUESTED** [review]
Looks good so far, please just take a look at the comments I left.

I especially like how you made the installation message (changelogs) stand out!

Thank you for your PR!


Though please run the linter and fix:
```
src/styles/options.scss
 231:1  ✖  Unexpected duplicate selector "body", first used at line 25   no-duplicate-selectors

src/styles/helpers/shared.scss
  44:9  ✖  Unexpected duplicate "transition"             declaration-block-no-duplicate-properties
 106:3  ✖  Expected selector ".enhancer-field label" to  no-descending-specificity                
           come before selector ".enhancer-field                                                  
           .enhancer-control label"                                                               
 106:3  ✖  Expected selector ".enhancer-field label" to  no-descending-specificity                
           come before selector ".enhancer-field                                                  
           .enhancer-control                                                                      
           input.enhancer-text-input:focus + label"                                               
 106:3  ✖  Expected selector ".enhancer-field label" to  no-descending-specificity                
           come before selector ".enhancer-field                                                  
           .enhancer-control                                                                      
           input.enhancer-text-input.has-value + label"                                           
 106:3  ✖  Expected selector ".enhancer-field label" to  no-descending-specificity                
           come before selector ".enhancer-field                                                  
           .enhancer-control                                                                      
           input.enhancer-text-input[placeholder] +                                               
           label"                                                                                 
 106:3  ✖  Expected selector ".enhancer-field label" to  no-descending-specificity                
           come before selector ".enhancer-field                                                  
           .enhancer-control                                                                      
           input.enhancer-text-input:disabled + label"
```

File: .gitignore (hunk 1 above)
Comment (groenroos) [reply .gitignore:5]: The biggest stumbling block for me initially was that it seems to need a very specific version of Node to work; I recall versions in both Node 12 and Node 16 not working. 

Node 12 doesn't like TypeScript;

```
[!] SyntaxError: Unexpected token '.'
/Users/groenroos/Repositories/NebulaEnhance/rollup.config.js:259
  const type = args.configType?.toLowerCase();
                               ^
```

Node 16.10 has some other problems:

```
Assertion failed: (thread_id_key != 0x7777), function find_thread_id_key, file ../src/coroutine.cc, line 134.
zsh: abort
```

I only managed to get it working with Node 14.x - which is fine, but would be good to at least document the version requirement in the development docs.

I also was never able to get the Firefox command running on my computer (macOS 11.4);

```
s: The request "dev" profile name cannot be resolved to a profile path
```

For this reason, I've only been able to test my changes on Chromium.

Comment (groenroos) [reply .gitignore:5]: Also, the `In depth` section in `DEVELOPING.md` mentions a `lint-fix` command which doesn't exist; and the `lint` command fixes problems automatically, which you wouldn't expect (at least based on the docs).

Comment (cpiber) [reply .gitignore:5]: I use Node version v15.14.0, that's very interesting.

> Node 12 doesn't like TypeScript;
This problem isn't TypeScript, the config is in ES; but it seems I am using some newer syntax, didn't catch that; will document.

> Node 16.10 has some other problems:
That, frankly, looks like a bug in node, that's an internal assertion, no clue how you got that.

> I also was never able to get the Firefox command running on my compute
I'm very sorry about that, it seems I forgot about this. In my setup, to make sure my settings (and passwords) aren't lost when I restart, I have a dedicated dev profile, I didn't notice that this is a problem, thank you for pointing it out!

Comment (cpiber) [reply .gitignore:5]: > Also, the In depth section in DEVELOPING.md mentions a lint-fix command which doesn't exist; and the lint command fixes problems automatically, which you wouldn't expect (at least based on the docs).

You're not up-to-date ;) I changed that recently

Comment (groenroos) [reply .gitignore:5]: I haven't been able to re-create the `build/` folder, but it only contained a `config.gypi` as per [this](https://stackoverflow.com/questions/18023501/what-is-this-build-config-gypi-file-and-can-i-delete-it). I recall having some kind of problems with `node-gyp` when setting up - perhaps related to installing `pnpm`, or to the various Node version issues I was troubleshooting before I thought to try Node 14?

Comment (cpiber) [reply .gitignore:5]: Hmm, I'm not sure, I don't see `node-gyp` anywhere in the lock file, so it's probably not anywhere in the dependencies...

Very strange, might be related to the error on Node 16

**Review by cpiber — COMMENTED** [review]
Thank you for your PR! Looks good!

**Comment by cpiber** [comment]
In case you are interested, I was just testing various versions of node, and stumbled upon the same problem you had with Node v16.

The problem is that `fibers` is not compatible with v16 and later: https://github.com/laverdet/node-fibers#readme. It is not deprecated and I have removed it in 70862e2

**Pull request merged**